### ISSUE TITLE
RISC-V LLVM feature +relax not recognized by rustc

### ISSUE THREAD

**Issue by zyedidia** [rust-lang/rust#109426]
### Code

```Rust
// Compile with: rustc --target riscv64imac-unknown-none-elf -C target-feature=+relax test.rs --emit=obj

#![no_std]
#![no_main]

use core::panic::PanicInfo;

#[panic_handler]
fn panic(_info: &PanicInfo) -> ! {
    loop {}
}
```

### Current output

```Shell
warning: unknown feature specified for `-Ctarget-feature`: `relax`
  |
  = note: it is still passed through to the codegen backend
  = help: consider filing a feature request

warning: 1 warning emitted
```

### Desired output

```Shell
No warning.
```

### Rationale and extra context

I would like to enable linker relaxation for the RISC-V target. To do so I passed `-C target-feature=+relax` to tell LLVM to use linker relaxation. It seems like `rustc` is not aware of this option, so it emits a warning (but the feature is still properly enabled). I can't find any way to disable this warning, so I am submitting this issue in the hopes that either the feature will become recognized, or a way to disable the warning will be added.

Related:  #96472.

The program is compiled by running `rustc --target riscv64imac-unknown-none-elf -C target-feature=+relax test.rs --emit=obj`

Hopefully I've posted this to the correct place. Thanks!

**Comment by zyedidia** [comment]
I tried it out and it does not seem to enable linker relaxation unfortunately.

**Comment by workingjubilee** [comment]
Why is a target feature being used for linker relaxation...?

**Comment by zyedidia** [comment]
This optimization/feature is specific to RISC-V. For the RISC-V target by default LLVM does not emit relocations (in the generated object file) for global variable lookups that tell the linker to make use of linker relaxation (generating addresses for globals using the `gp` register). I’m not sure of the rationale behind this (gcc enables it by default and uses `-mno-relax` to disable it). One reason might be that using linker relaxation does require the runtime or something to set up the `gp` register beforehand.

In any case it should be possible to enable with Rust without getting a warning, and perhaps should even be enabled by default for RISC-V targets.

**Comment by workingjubilee** [comment]
As mentioned in the other issue, the warning is in large part because LLVM and Rust don't always agree on the definition of a target feature, in part because LLVM has proven surprisingly willing to change the definitions.
